src/api_client.ts

```diff
@@ -0,0 +1,89 @@
+import restClient, {Options, RequestPromise} from 'request-promise';
+import { createGovAuNotifyToken } from './authentication';
+import packageJson from '../package.json';
+
+export default class NotifyClient {
+
+  proxy: string;
+  baseUrl: string = 'https://rest-api.notify.gov.au';
+  secret: string;
+  serviceId: string;
+
+  constructor(params: {apiKeyId: string, baseUrl?: string}) {
+
+    if (params.baseUrl) {
+      this.baseUrl = params.baseUrl;
+    }
+
+    this.serviceId = this.getServiceIdFromApiKey(params.apiKeyId);
+    this.secret = this.getRealApiKeyFromApiKey(params.apiKeyId);
+  }
+
+
+  /**
+   * @param {string} path
+   *
+   * @returns {Promise}
+   */
+  public get(path: string): RequestPromise {
+    const options : Options = {
+      method: 'GET',
+      uri: this.baseUrl + path,
+      json: true,
+      resolveWithFullResponse: true,
+      headers: this.getHeader()
+    };
+
+    if(this.proxy !== null) {
+      options.proxy = this.proxy;
+    }
+
+    return restClient(options);
+  }
+
+  /**
+   * @param {string} path
+   * @param {object} data
+   *
+   * @returns {Promise}
+   */
+  public post(path: string, data?: object): RequestPromise {
+    const options : Options = {
+      method: 'POST',
+      uri: this.baseUrl + path,
+      json: true,
+      body: data,
+      resolveWithFullResponse: true,
+      headers: this.getHeader()
+    };
+
+    if(this.proxy !== null) {
+      options.proxy = this.proxy;
+    }
+
+    return restClient(options);
+  }
+
+  /**
+   * @param {String} url
+   */
+  public setProxy(url: string) {
+    this.proxy = url;
+  }
+
+  private getHeader(): object {
+    return {
+      'Authorization': 'Bearer ' + createGovAuNotifyToken(this.secret, this.serviceId),
+      'User-agent': 'NOTIFY-API-NODE-CLIENT/' + packageJson.version
+    };
+  }
+
+  private getServiceIdFromApiKey(apiKeyId: string): string {
```

Comment:
Can just be `serviceIdFromApiKey` and doesn't have to be on the class. Move it to the module top level—easier to test.

Reply:
Out of curiosity how is it easier to test? via libraries like `rewire`?

Reply:
Because `getServiceIdFromApiKey` doesn't use anything from inside the class (`this.whatever`), it doesn't need to be a class method. It's arguably easier to test because your `api_client.test.ts` can just directly `import { getServiceIdFromApiKey } from './api_client'` and test it rather than having to instantiate an instance of the class.